src/vision/status_quo/barbara_bridges_sync_and_async.md

```diff
@@ -0,0 +1,245 @@
+```
+
+Of course! `spawn_blocking`, like `map`, just takes a regular closure, not an async closure. She's getting a bit frustrated now. "Well," she thinks, "I can use `spawn` to get into an async context!" So she adds a call to `spawn` inside the `spawn_blocking` closure:
+
+```rust
+fn aggregate(urls: &[Url]) {
+    let data: Vec<Data> =
+        urls
+            .iter()
+            .map(|url| tokio::task::spawn_blocking(move || {
+                tokio::task::spawn(async move {
+                    do_web_request(url).await
+                })
+            }))
+            .collect();
+}
+```
+
+But this isn't really helping, as `spawn` still yields a future. She's getting the same errors.
+
+### Async all the way
+
+She remembers now that this whole drama started because she was converting her `main` function to be `async`. Maybe she doesn't have to bridge between sync and async? She starts digging around in the docs and finds `futures::join_all`. Using that, she can change `aggregate` to be an async function too:
+
+```rust
+async fn aggregate(urls: &[Url]) {
+    let data: Vec<Data> = futures::join_all(
+        urls
+            .iter()
+            .map(|url| do_web_request(url))
+    ).await;
+}
+```
+
+Things are working again now, so she is happy.
+
+### Filtering
+
+Later on, she would like to apply a filter to the aggregation operation. She realizes that if she wants to use the fetched data when doing the filtering, she has to filter the vector after the join has completed; `join_all` doesn't have a way to put a filter into the iterator chain like she wants. She is annoyed, but performance isn't critical, so it's ok.
+
+### And the cycle begins again
+
+Later on, she wants to call `aggregate` from another binary. This one doesn't have an `async main`. This context is deep inside of an iterator chain and was previously entirely synchronous. She realizes it would be a lot of work to change all the intervening stack frames to be `async fn`, rewrite the iterators into streams, etc. She decides to just call `block_on` again, even though it make her nervous.
+
+## 🤔 Frequently Asked Questions
+
+### What are the morals of the story?
+
+* Some projects don't care about max performance and just want things to work.
+    * They would probably be happy with sync but as the most popular libraries for web requests, databases, etc, offer async interfaces, they may still be using async code.
+* There are contexts where you can't easily add an `await`.
+    * For example, inside of an iterator chain.
+    * Big block of existing code.
+* Mixing sync and async code (`block_on`) can cause deadlocks that are really painful to diagnose.
+
+
+### Why did you choose Barbara to tell this story?
+
+* Because Mark (who experienced most of it) is a very experienced Rust developer.
+* Because you could experience this story regardless of language background or being new to Rust.
```

Comment:
Just thinking out loud here but a bit unsure Barbara would not rethink their use of sync iterator at the first sign of trouble, especially at the point where they change `aggregator` to be an async function. Surely, they'd know they need async iterators (Streams) and very likely know of APIs like [`unfold`](https://docs.rs/futures/0.3.14/futures/stream/fn.unfold.html) that can be used to create streams out of futures. Streams would also solve her filtering requirement. Maybe Alan would fit better here?

Reply:
No, even Barbara who is an experienced rust developer but new to async can still trip here. `Stream`s are all new things in async, when I got into it I am also a bit confused at first, I was wondering if `Stream` is exactly the same as `Iterator` which one item is done before another or if the items processed can interleave (and indeed it can interleave), even when I read the `Stream` docs multiple times.

Reply:
> I was wondering if `Stream` is exactly the same as `Iterator`

Maybe you didn't go through **the** [async book](https://rust-lang.github.io/async-book/05_streams/01_chapter.html), which AFAIK we've been assuming our characters go through (especially Barbara) and it introduces streams as such from the very beginning:

> The Stream trait is similar to Future but can yield multiple values before completing, similar to the Iterator trait from the standard library

and

> Similar to synchronous Iterators, there are many different ways to iterate over and process the values in a Stream. There are combinator-style methods such as map, filter, and fold, and their early-exit-on-error cousins try_map, try_filter, and try_fold.

but as I said, I was just thinking out loud.

Reply:
Sorry, I didn't go through the async book (actually I did but the time I read it there isn't `Stream` in the book). But even in the characters set, Barbara wasn't said to go through async book, it just says

> She is using Rust for its feeling of productivity, and she expects Async Rust to continue in that tradition.

> She wants a design that feels like the rest of Rust.

> She loves Rust and she expects Async Rust to share its overall values.

So I believe Barbara may not go through async book given that she thinks it behaves similarly.

Reply:
> But even in the characters set, Barbara wasn't said to go through async book

Oh, I didn't mean to imply that it's a declared assumption but during the meetings, that's what I understood was the assumption. @nikomatsakis what do you say?

Reply:
I am not assuming folks go through the async book, just that Barbara is experienced with Rust but new to async. I would hazard a guess that few people sit down to read the async book -- most people I've talked to don't know it exists!